File: app/src/at-sonnguyen/java/com/asiantech/intern20summer1/Intro/RainsOfReason.java
package com.asiantech.intern20summer1.Intro;

public class RainsOfReason {
    /*
    * replace all the occurrences of elemToReplace with substitutionElem
    */
    static int[] arrayReplace(int[] inputArray, int elemToReplace, int substitutionElem) {
        for(int index=0;index<inputArray.length;index++) {
            if(inputArray[index]==elemToReplace)
                inputArray[index]=substitutionElem;
        }
        return inputArray;
    }
    /*
    *Check if all digits of the given integer are even.
    */
    static boolean evenDigitsOnly(int n) {
        while ( n > 0 ){
            if (n%2 != 0 ){
                return false;
            }
            n /= 10 ;
        }
        return true ;
    }
    /*
    * Check  variable name
    */
    static boolean variableName(String name) {
        if (!((name.charAt(0) >= 'a' && name.charAt(0) <= 'z')
                || (name.charAt(0)>= 'A' && name.charAt(0) <= 'Z')
                || name.charAt(0) == '_'))
            return false;
        for (int i = 1; i < name.length(); i++)
        {
            if (!((name.charAt(i) >= 'a' && name.charAt(i) <= 'z')
                    || (name.charAt(i) >= 'A' && name.charAt(i) <= 'Z')
                    || (name.charAt(i) >= '0' && name.charAt(i) <= '9')
                    || name.charAt(i) == '_'))
                return false;
        }
        return true;
    }
    /*
    * eplace each of its characters by the next one
    */
    static String alphabeticShift(String inputString) {
        char[] inputStringArray = inputString.toCharArray();
        for (int i = 0; i < inputStringArray.length; i++) {
            inputStringArray[i]++;
            if (inputStringArray[i] > 'z')
                inputStringArray[i] = 'a';
        }
        return new String(inputStringArray);
    }
    /*
    *Given two cells on the standard chess board, determine whether they have the same color or not.
    */
    static boolean chessBoardCellColor(String cell1, String cell2) {
        if((cell1.charAt(0) + cell1.charAt(1)) % 2 == 0 &&
                (cell2.charAt(0) + cell2.charAt(1)) % 2 == 0 ||
                (cell1.charAt(0) + cell1.charAt(1)) % 2 != 0 &&
                        (cell2.charAt(0) + cell2.charAt(1)) % 2 != 0) {
            return true;

        }
        return false;
    }

    public static void main(){
        int[] inputArray = {1,2,3,4,1,5,1};
        int[] ResultArray = arrayReplace(inputArray , 1,3);
        for ( int i = 0 ; i< ResultArray.length ; i++){
            System.out.print(ResultArray[i]);
        }
        System.out.print("Check EvenDigitsOnly: "+evenDigitsOnly(168));
        String StringToCheck = "variable___to_check";
        System.out.print( "Result Check VariableName"+variableName(StringToCheck));
        String StringToReplace = "abcaz";
        String CheckReplaceChar = alphabeticShift(StringToReplace);
        System.out.print("after Replace" +CheckReplaceChar);
        System.out.print("Same Color: "+chessBoardCellColor("A1","C3"));


    }
}

File: app/src/at-sonnguyen/java/com/asiantech/intern20summer1/kotlin/RainsOfReason.kt
package com.asiantech.intern20summer1.kotlin

object RainsOfReason{
    private const val TEN = 10
    /*
    *   Given an array of integers, replace all the occurrences of elemToReplace with substitutionElem.
    */
    private fun arrayReplace(inputArray: MutableList<Int>, elemToReplace: Int, substitutionElem: Int): MutableList<Int> {
        for(i in inputArray.indices){
            if (inputArray[i] == elemToReplace ){
                inputArray[i] = substitutionElem
            }
        }

        return inputArray
    }
    /*
    * Check if all digits of the given integer are even.
    */
    private fun evenDigitsOnly(n: Int): Boolean {
        var copyN : Int = n
        while(copyN > 0){
            if(copyN % 2 != 0 ){
                return false
            }
            copyN/= TEN
        }
        return true
    }
    /*
    *   Check if the given string is a correct variable name.
    */
    private fun variableName(name: String): Boolean {
        return """([a-zA-Z]|_)+(\w|_)*""".toRegex().matches(name)
    }
    /*
    *   Given a string, your task is to replace each of its characters
    *   by the next one in the English alphabet
    */
    private fun alphabeticShift(inputString: String): String {
        var resultString : String = ""
        for (i in inputString.indices){

            if (inputString.get(i)+1 > 'z'){
                resultString += "a"
            }else resultString += (inputString.get(i)+1)
        }
        return resultString
    }
    /*
    *   Given two cells on the standard chess board, determine whether they have the same color or not.
    */
    private fun chessBoardCellColor(cell1: String, cell2: String): Boolean =
        (cell1[0].toInt()+cell1[1].toInt())%2 == (cell2[0].toInt()+cell2[1].toInt())%2
    @JvmStatic
    fun main(args: Array<String>) {
        val inputArray : MutableList<Int> = mutableListOf(1, 2, 1)
        //val n : Int = 268462
        val string = "variable0"
        val cell1 = "A1"
        val cell2 = "H2"
        println("25. Replace all element = 1 to element = 3 in $inputArray is ${arrayReplace(inputArray,1,3)} ")
        println("26.  is only even Digits ? ${evenDigitsOnly(268462)} ")
        println("27. $string is a variable name ? ${variableName(string)} ")
        println("28. $string after shift a value is ${alphabeticShift(string)}")
        println("29. $cell1 and $cell2 are same color ? ${chessBoardCellColor(cell1,cell2)} ")
    }
}
